src/librustc_mir/const_eval.rs

```diff
@@ -520,25 +503,41 @@ fn validate_and_turn_into_const<'a, 'tcx>(
     constant: RawConst<'tcx>,
     key: ty::ParamEnvAnd<'tcx, GlobalId<'tcx>>,
 ) -> ::rustc::mir::interpret::ConstEvalResult<'tcx> {
-    let cid = key.value;
     let ecx = mk_eval_cx(tcx, tcx.def_span(key.value.instance.def_id()), key.param_env);
     let val = (|| {
-        let op = ecx.raw_const_to_mplace(constant)?.into();
-        // FIXME: Once the visitor infrastructure landed, change validation to
-        // work directly on `MPlaceTy`.
-        let mut ref_tracking = RefTracking::new(op);
-        while let Some((op, path)) = ref_tracking.todo.pop() {
+        let mplace = ecx.raw_const_to_mplace(constant)?;
+        let mut ref_tracking = RefTracking::new(mplace);
+        while let Some((mplace, path)) = ref_tracking.todo.pop() {
             ecx.validate_operand(
-                op,
+                mplace.into(),
                 path,
                 Some(&mut ref_tracking),
                 true, // const mode
             )?;
         }
         // Now that we validated, turn this into a proper constant.
-        let def_id = cid.instance.def.def_id();
-        let normalize = tcx.is_static(def_id).is_none() && cid.promoted.is_none();
-        op_to_const(&ecx, op, normalize)
+
+        // We also store a simpler version of certain constants in the `val` field of `ty::Const`
+        // This helps us reduce the effort required to access e.g. the `usize` constant value for
+        // array lengths. Since array lengths make up a non-insignificant amount of all of the
+        // constants in the compiler, this caching has a very noticeable effect.
+
+        // FIXME(oli-obk): see if creating a query to go from an `Allocation` + offset to a
+        // `ConstValue` is just as effective as proactively generating the `ConstValue`.
+        let val = match mplace.layout.abi {
+            layout::Abi::Scalar(..) => {
+                let scalar = ecx.try_read_immediate_from_mplace(mplace)?.unwrap().to_scalar()?;
+                ConstValue::Scalar(scalar)
+            }
+            layout::Abi::ScalarPair(..) if mplace.layout.ty.is_slice() => {
+                let (a, b) = ecx.try_read_immediate_from_mplace(mplace)?.unwrap().to_scalar_pair()?;
+                ConstValue::Slice(a, b.to_usize(&ecx)?)
+            },
+            _ => ConstValue::ByRef,
+        };
+        let ptr = Pointer::from(constant.alloc_id);
+        let alloc = constant.alloc;
+        Ok(ty::Const { val, ty: mplace.layout.ty, alloc: Some((alloc, ptr))})
```

Comment:
The code above here is basically `mplace_to_const`, right? And we have the same in `const_field`, don't we? Couldn't that become a helper function?